### ISSUE TITLE
for

### ISSUE THREAD

**Issue by tebe6502** [KarolS/millfork#8]
OK, if constants
for i,1,to,0

WRONG, if variable
j=1
for i,j,to,0

**Comment by KarolS** [comment]
That's a pretty ugly inconsistency, thanks for pointing it out.

The code generator has special cases for certain loops with constant bounds and one of those cases is `for i,s,to,e` loop where `s` and `e` are known constants and `s > e`, which is rewritten to a no-op. But, as you noticed, it's inconsistent with how non-constant bounds with the same values work.

After thinking about it for a while, I decided that performance is the most important thing and checking if the bounds are in a wrong order is too expensive. Usually they are not, so if the programmer really doesn't want to deal with overflow, then they have to check it themself.

I'm now working on making the second behaviour the default guaranteed one. I've removed all special cases that are inconsistent with everything else and I'm currently looking through various pieces of old Millfork code to see if this change breaks anything, but I don't expect any major problems.

Also, this will make it possible to do things like easily iterating over non-printable ASCII characters: `for c,$7F,to,$1F`
